linera-service/src/client.rs

```diff
@@ -0,0 +1,1002 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::config::WalletState;
+use anyhow::{Context, Result};
+use async_graphql::InputType;
+use linera_base::{
+    abi::ContractAbi,
+    identifiers::{ChainId, MessageId, Owner},
+};
+use linera_execution::Bytecode;
+use once_cell::sync::OnceCell;
+use serde::ser::Serialize;
+use serde_json::{json, value::Value};
+use std::{
+    collections::{BTreeMap, HashMap, HashSet},
+    env, fs,
+    ops::RangeInclusive,
+    path::{Path, PathBuf},
+    process::Stdio,
+    rc::Rc,
+    str::FromStr,
+    time::Duration,
+};
+use tempfile::{tempdir, TempDir};
+use tokio::{
+    process::{Child, Command},
+    sync::Mutex,
+};
+use tonic_health::proto::{
+    health_check_response::ServingStatus, health_client::HealthClient, HealthCheckRequest,
+};
+use tracing::{info, warn};
+
+/// The name of the environment variable that allows specifying additional arguments to be passed
+/// to `cargo` when starting client, server and proxy processes.
+pub const CARGO_ENV: &str = "INTEGRATION_TEST_CARGO_PARAMS";
+
+/// The name of the environment variable that allows specifying additional arguments to be passed
+/// to the binary when starting a server.
+const SERVER_ENV: &str = "INTEGRATION_TEST_SERVER_PARAMS";
+
+/// The name of the environment variable that allows specifying additional arguments to be passed
+/// to the node-service command of the client.
+const CLIENT_SERVICE_ENV: &str = "INTEGRATION_TEST_CLIENT_SERVICE_PARAMS";
+
+#[derive(Copy, Clone)]
+pub enum Network {
+    Grpc,
+    Simple,
+}
+
+impl Network {
+    fn internal(&self) -> &'static str {
+        match self {
+            Network::Grpc => "\"Grpc\"",
+            Network::Simple => "{ Simple = \"Tcp\" }",
+        }
+    }
+
+    fn external(&self) -> &'static str {
+        match self {
+            Network::Grpc => "\"Grpc\"",
+            Network::Simple => "{ Simple = \"Tcp\" }",
+        }
+    }
+
+    fn external_short(&self) -> &'static str {
+        match self {
+            Network::Grpc => "grpc",
+            Network::Simple => "tcp",
+        }
+    }
+}
+
+pub struct Client {
+    tmp_dir: Rc<TempDir>,
+    storage: String,
+    wallet: String,
+    max_pending_messages: usize,
+    network: Network,
+}
+
+impl Client {
+    fn new(tmp_dir: Rc<TempDir>, network: Network, id: usize) -> Self {
+        Self {
+            tmp_dir,
+            storage: format!("rocksdb:client_{}.db", id),
+            wallet: format!("wallet_{}.json", id),
+            max_pending_messages: 10_000,
+            network,
+        }
+    }
+
+    pub async fn project_new(&self, project_name: &str) -> Result<TempDir> {
```

Comment:
Most of these methods aren't used in production. We could put them into a `#[cfg(any(test, feature = "test"))] impl Client`. Or even move them back into the end-to-end test module, as part of a `ClientExt` trait or something?

Reply:
Yes that's a good point. I think I want to land this now so I'll make a new issue for this and address it in a subsequent PR. Might also be a good first issue for newcomers to the project.